Docs/BazelForiOSDevelopers.md

```diff
@@ -0,0 +1,522 @@
+
+_- [the bazel documentation](https://docs.bazel.build/versions/master/skylark/rules.html)_
+
+
+In the previous segment, we created an iOS application with via a BUILD file.
+The `ios_application` rule is implemented by `rules_apple` and the
+`objc_library` is a native rule. Native rules are implemented in Java and have
+more power and capabilities needed to implement core functionality like remote
+execution and c++ compilation and are generally used as-is.
+
+```
+BUILD file -> target -> rule -> action -> execution
+```
+
+Bazel and open source rules should provide most functionality to build an iOS
+application. Generally, defining custom rules isn't required but can improve and
+consolidate functionality. To learn more about how to create custom rules, the
+document, [Bazel
+extensions](https://docs.bazel.build/versions/master/skylark/concepts.html)
+contains a comprehensive overview.
+
+### Macros
+
+Like [rules](#Rules) and [aspects](#Aspects), macros are
+defined in `.bzl` files.  A macro is a convenient way to call a rule, and not
+recognized by Bazel in the same way a rule is.
+
+_Note: The main distinction between a `.bzl` and a `BUILD` file is `BUILD` files are
+used to create targets by calling macros and rules. `.bzl` files define the
+implementation._
+
+Most users of Bazel implement a higher level system of macros to encapsulate
+defaults of building librarys and simplify configuration management. To create
+a wrapper for `objc_library`, create the file `objc_library.bzl`. The following
+macro restricts the customization, and enforces defaults of the native
+`objc_library` rule.
+
+```
+def objc_library(name, srcs=[], hdrs=[], deps=[], data=[]):
+    """
+    An objc_library that turns on pedantic warnings,
+    and enables modules
+    """
+    native.objc_library(
+        name=name,
+        deps=deps,
+        data=data,
+        srcs=srcs,
+        hdrs=hdrs,
+        enable_modules=True,
+        copts=[
+            "-Wpedantic"
+        ]
+    )
+```
+
+By loading our `objc_library` into a `BUILD` file, it will override the native
+`objc_library` rule, which is automatically imported.
+```
+load(":objc_library.bzl", "objc_library")
+
+# This calls the custom objc_library macro. Thankfully, building this with
+# Bazel will error out due to the unavailable argument `copts` used here.
+objc_library(name="some", copts=["-DSOME"])
+```
+
+The same principals can be applied to many rules: wrapping macros with macros.
+Please see the
+[`objc_library`](https://docs.bazel.build/versions/master/be/objective-c.html)
+documentation for all possible arguments. _Note: Unlike Starlark rules which
+are easily added on to Bazel, the `objc_library` is part of the internal java
+rules shipped with the Bazel
+binary._ 
+
+_In the Xcode world, configuration is defined within Xcode projects and
+proprietary xcconfigs. In Bazel it's easy to establish norms, enforce
+consistency with abstractions like `objc_library`, and refactor_
```

Comment:
```suggestion
In the Xcode world, configuration is defined within Xcode projects and
proprietary xcconfigs. In Bazel it's easy to establish norms, enforce
consistency with abstractions like `objc_library`, and safely refactor.
```

1. This doesn't seem to fit the mold of other places using italics in this doc.
2. Is "machine-readable" better than "proprietary" here?

Reply:
Good point, perhaps we can have mention this in depth some where and how Bazel is fundamentally different than what the point of an xcconfig is:
- xcconfigs are a makefile like syntax to configure Xcode build behavior
- Xcode interprets xcconfigs to peform actions in a way that is opaque and unmodifiable by the user
- bazel doesn't have xcconfigs and puts the user in control of the build through settings, rules, aspects, toolchains etc